### PR TITLE
🍒[5.7][Distributed] Fixes to distributed funcs and protocols

### PR DESCRIPTION
**Pull request by ktoso**
Cherry pick of:
- #59711
- #59718 
- and soon the step 3 as well

Resolves:
- rdar://95829405
-  rdar://92757561

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please test

File: lib/SILGen/SILGenPoly.cpp [hunk 1]
```diff
@@ -4435,15 +4435,22 @@ void SILGenFunction::emitProtocolWitness(
...
+      getActorIsolation(witness.getDecl()).isDistributedActor()) {
+    // We witness protocol requirements using the distributed thunk, when:
+    // - the witness is isolated to a distributed actor, but the requirement is not
+    // - the requirement is a distributed func, and therefore can only be witnessed
+    //   by a distributed func; we handle this by witnessing the requirement with the thunk
+    // FIXME(distributed): this limits us to only allow distributed explicitly throwing async requirements... we need to fix this somehow.
+    if (requirement.hasDecl()) {
+      if ((!getActorIsolation(requirement.getDecl()).isDistributedActor()) ||
+          (isa<FuncDecl>(requirement.getDecl()) &&
+              witness.getFuncDecl()->isDistributed())) {
```
Comment (DougGregor) [inline lib/SILGen/SILGenPoly.cpp:4448]: I'd prefer if the fix from https://github.com/apple/swift/pull/59722 were pulled into this PR, so we're in a reasonable resting state for 5.7.
```suggestion
              requirement.getFuncDecl()->isDistributed())) {
```

Comment (ktoso) [reply lib/SILGen/SILGenPoly.cpp:4448]: okey, will do

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please test

**Review by DougGregor — APPROVED** [review]

**Pull request merged**